### ISSUE TITLE
[REVIEW]: BCAW: Automated tool for codon usage bias analysis for molecular evolution

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#1500]
**Submitting author:** @AliYoussef96 (<a href="http://orcid.org/0000-0002-5201-387X">Ali Youssef</a>)
**Repository:** <a href="https://github.com/AliYoussef96/BCAW-Tool" target ="_blank">https://github.com/AliYoussef96/BCAW-Tool</a>
**Version:** v1.0.3
**Editor:** @mgymrek
**Reviewer:** @Benjamin-Lee
**Archive:** <a href="https://doi.org/10.5281/zenodo.3468557" target="_blank">10.5281/zenodo.3468557</a>

## Status

[![status](http://joss.theoj.org/papers/5c17f813c2eca6b9d7c4ecf5d2ea97e9/status.svg)](http://joss.theoj.org/papers/5c17f813c2eca6b9d7c4ecf5d2ea97e9)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/5c17f813c2eca6b9d7c4ecf5d2ea97e9"><img src="http://joss.theoj.org/papers/5c17f813c2eca6b9d7c4ecf5d2ea97e9/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/5c17f813c2eca6b9d7c4ecf5d2ea97e9/status.svg)](http://joss.theoj.org/papers/5c17f813c2eca6b9d7c4ecf5d2ea97e9)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/AliYoussef96/BCAW-Tool" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@Benjamin-Lee, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @mgymrek know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/AliYoussef96/BCAW-Tool">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** v1.0.3
- [x] **Authorship:** Has the submitting author (@AliYoussef96) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @Benjamin-Lee it looks like you're currently assigned to review this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01500/joss.01500/10.21105.joss.01500.pdf)

**Comment by AliYoussef96** [comment]
Kindly I'm asking when the reviewing process will start.

**Comment by kyleniemeyer** [comment]
Hello @Benjamin-Lee, any updates on your review, or when you expect to complete it? Thanks!

**Comment by Benjamin-Lee** [comment]
Sorry, I've been swamped with travel. I'll start my review now.

**Comment by Benjamin-Lee** [comment]
Ok, I just added five initial issues to the repo. @AliYoussef96, I hope my issues are understandable but if not, please don't hesitate to reach out for more information. I'm excited to try this tool out!

**Comment by mgymrek** [comment]
Hi @AliYoussef96

I agree with the points raised by @Benjamin-Lee. See also my comments below.

### Install process
I agree this should be structured more like a standard Python library with a setup.py file for easy installation of the tool and dependencies.

### Usage/docs
1. I personally prefer that tools take in all the arguments at once rather than prompting for each one. That way it can be more easily embedded in other pipelines that might not directly interact with uses. You could also give a complete command on example data for someone to try.
2. It would be helpful if the README specified in the initial description what are the inputs and outputs. It looks like the input is a fasta file with gene sequences, rather than just any fasta file (e.g. with a reference genome). It could also be helpful to point users to how to obtain such fasta files for a species of interest. 
3. I could not find documentation of what the plots are showing. e.g. what do GC1, GC2, GC3 mean for the first plot shown and what do Axis 1 vs. Axis 2 mean in the second plot?

### README and paper:
REAMDE: "automated workflow to analysis the codon usage " -> "automated workflow to analyze the codon usage'
There are now references in the paper. It would also be helpful for the paper to describe in more detail what the outputs are.

### Testing
I do not see tests. It would be great to both tests functions with automated tests, and also to provide some indication that your overall results are correct. e.g. by comparing E. Coli results to previously published ones.

**Comment by arfon** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01500/joss.01500/10.21105.joss.01500.pdf)

**Comment by AliYoussef96** [comment]
Hi @mgymrek
 
**Install process**

For [setup.py](https://github.com/AliYoussef96/BCAW-Tool/blob/master/setup.py) is available now, as well as installation using pip.

```
pip install BCAWT
```
**Usage/docs**

1- BCAWT can now take all the arguments at once.

```
from BCAWT import BCAWT
BCAWT.BCAW('Ecoli','result_folder',genetic_code_=11,fasta=True,Auto=True)
```
2- Inputs, inputs, and how to obtain such fasta files are now on README.
As well as I added [Abbreviations table](https://github.com/AliYoussef96/BCAW-Tool/blob/master/Abbreviations.md), Also, I added an introduction to codon usage bias and equations used in the BCAW tool with some useful references [here](https://github.com/AliYoussef96/BCAW-Tool/blob/master/Introduction%20to%20codon%20usage%20bias.pdf).

3- All plots are described [here](https://github.com/AliYoussef96/BCAW-Tool/blob/master/Plots/Plots%20Description.md). 

**Testing**

For the BCAW tool, I did BCAWT_auto_test.py. I did describe how to use this test in README.md and in the documentation.

Also, I wrote [test.py](https://github.com/AliYoussef96/BCAW-Tool/blob/master/test.py) to tests functions with automated tests

**paper**

I am working on a summary to describe in more detail what the outputs are.

**paper**

**Comment by AliYoussef96** [comment]
Hi @mgymrek 

For the [ paper](https://github.com/AliYoussef96/BCAW-Tool/blob/master/paper/paper.md) Output summary, I add a table summary, as well as a figure, shows all the output plots.

I'm Waiting for your feedback.

Hi @Benjamin-Lee 

I finished working on the issues you added. And I'm Waiting for your feedback.

**Comment by mgymrek** [comment]
thanks @AliYoussef96 for the updates. The changes look good.

One comment on install: I wonder whether it is critical to require Python >=3.7? Many systems will still have earlier versions making it difficult to install and test with pip.

Also, comments on the paper:

Summary
* "Futher, Using" -> "Further, using the"
* "The tool also includes statistical analysis as" -> "The tool also includes statistical analysis such as"

Implementation
* "build in" -> "built in"
* "bunch of analysis" -> "variety of analyses"
* It is not as useful to say how many different files are output. It would more helpful to just summarize the types of output. e.g. plots showing XYZ
* "BCAWtool" -> "BCAW Tool". same elsewhere, make sure you are consistent with how you reference the tool

Output summary:
* "CSV files contain" -> "CSV file containing"
* Start a new sentence after Table 1, e.g. "It also returns 11 plots depicting output for easy interpretation (Fig. 1)"

**Comment by AliYoussef96** [comment]
Thanks, @mgymrek for your comments.

**1. Comment on install to require Python >=3.7**.
[(README)](https://github.com/AliYoussef96/BCAW-Tool/blob/master/README.md).

**2. Updates to the paper**

**Summary**

- The summary section is updated.

**Implementation**

- The Implementation section is updated. For "BCAWtool" I converted it to "BCAW tool" everywhere in the paper and also in the readme.

**Output summary**

- The Output summary section is updated.

Waiting for further comments.

**Comment by AliYoussef96** [comment]
Hi @Benjamin-Lee 

Can we close the opened issues now?

**Comment by labarba** [comment]
@mgymrek — it looks like the author is waiting for a response here: what's the status of this submission?

**Comment by mgymrek** [comment]
Thanks for the updates. @Benjamin-Lee can you also finish going throug the checklist after these revisions to see if remaining requirements are met?

One point: the pip version is 0.2.0 but the version listed on this review is v1.0.0. Which one is the most recent release?

**Comment by AliYoussef96** [comment]
I updated the pip version to 1.0.0. to give the match with the GitHub release (v1.0.0).

[*The pip version*](https://pypi.org/project/BCAWT/)

**Comment by Benjamin-Lee** [comment]
I’ll get on it!

**Comment by AliYoussef96** [comment]
@Benjamin-Lee and @mgymrek 

About the checklist, any comments about the Functionality and the References?

**Comment by Benjamin-Lee** [comment]
Not yet. Please be aware that I am on vacation starting out of internet service, so I will get back to you in a week!

**Comment by mgymrek** [comment]
Hi @Benjamin-Lee just wanted to check in on this review. Thanks!

**Comment by mgymrek** [comment]
@AliYoussef96 I apologize for the delay in moving this forward. I have been waiting to hear back from the reviewer. Regardless of that I will update you with a decision by the end of this week.

**Comment by Benjamin-Lee** [comment]
Hi! I’ll get on this tomorrow. Sorry for the delay.

**Comment by AliYoussef96** [comment]
> Just worked on trying it out and am currently running into an error on Mac. Once I can review the functionality, we'll be good to go.

Hi @Benjamin-Lee  

I made changes in the main function and I believe this will solve the error. 

https://github.com/AliYoussef96/BCAW-Tool/issues/12#issuecomment-531560160

**Comment by mgymrek** [comment]
Hello,
I am testing out these changes. Using the version from pip and the example:

```from BCAWT import BCAWT_auto_test
path = "Test_folder" # or any!!
BCAWT_auto_test.auto_test(path)
#processing....
BCAWT_auto_test.auto_check_files(path)
```

I get an error:
```
Bio.Data.CodonTable.TranslationError: Codon '<DI' is invalid
```

The main example (`BCAWT.BCAW(['Ecoli.fasta'],'save_path',genetic_code_=11,Auto=True)`) worked smoothly.

**Comment by AliYoussef96** [comment]
Hello @mgymrek,

The error you got from biopython translate function when it tries to translate unknown codons.  Any invalid codon (e.g. "TA?" or "T-A") will throw a TranslationError.
link to the error description : https://biopython.org/DIST/docs/api/Bio.Seq-module.html#translate

According to the error, a codon "**<DI**" is in one of DNA sequences in the fasta file being tested, I'm sure the [test file](https://github.com/AliYoussef96/BCAW-Tool/blob/master/tests/Ecoli.fasta) I added on the repository has valid DNA sequences. 

The automatic testing you trying to run expect this file.
```python
from BCAWT import BCAWT_auto_test
path = "Test_folder" # or any!!
BCAWT_auto_test.auto_test(path)
#processing....
BCAWT_auto_test.auto_check_files(path)
```
So, kindly can you make sure you have the write file.

**Comment by mgymrek** [comment]
Ok, I got it to work for now. The usage on the site is misleading. I assumed "Test_folder" is a folder containing the fasta files to check. (The functionality seemed to change since I tried a couple days ago using the pip version, since earlier I gave a directory and it didn't complain that it couldn't find the Ecoli.fasta file)
The README mentions any fasta file can be downloaded under "Auto testing". However "Ecoli.fasta" appears to be hardcoded into the test function. The usage should clarify whether that specific file is needed. If it is, that should be part of the default install so that users don't have to seprately fetch that file to test.


Also, here are some comments on the paper:
"have been developed to analysis" -> "have been developed to analyze"
"of any organism genes" -> "of any organism's genes"
"Further, Using" -> "Further, using"
"CUB as;" -> "CUB such as;"
"statistical analysis as" -> "statistical analysis such as"
"with build in" -> "with built in"
"a bunch of analysis" -> "a variety of analyses"

Once the above are taken care of we should be ready to go.

**Comment by AliYoussef96** [comment]
1- I updated your comments on the paper.

2- I changed the test function to accept any fasta file and I made it clear in the README.

```python
from BCAWT import BCAWT_auto_test
path = "Test_folder" # absolute path to the directory to save the result in
test_fasta = "Test_fasta_file" # absolute path to the fasta file that will be tested 
BCAWT_auto_test.auto_test(path, test_fasta)
#processing....
BCAWT_auto_test.auto_check_files(path) # note: this test assumes that in the result folder nothing except the result files form the above function.
```
3- The pip version now it 1.0.3.

**Comment by mgymrek** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #1500 with the following error: 

 [WARNING] Could not convert image '/tmp/tex2pdf.974/0872d19f51538b8f623dfc4af45d7772d82b17cf.shtml': Cannot load file
  Jpeg Invalid marker used
  PNG Invalid PNG file, signature broken
  Bitmap Invalid Bitmap magic identifier
  GIF Invalid Gif signature : 
  
  
  
  
  
  
  HDR Invalid radiance file signature
  Tiff Invalid endian tag value
  TGA Invalid bit depth (104)
Error producing PDF.
! LaTeX Error: Cannot determine size of graphic in /tmp/tex2pdf.974/0872d19f515
38b8f623dfc4af45d7772d82b17cf.shtml (no BoundingBox).

See the LaTeX manual or LaTeX Companion for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
l.476 ...9f51538b8f623dfc4af45d7772d82b17cf.shtml}

Looks like we failed to compile the PDF

**Comment by mgymrek** [comment]
@labarba do you have any insights into why this proof is failing to generate?

**Comment by danielskatz** [comment]
You might try changing the path to the image in the .md file to 

`https://github.com/AliYoussef96/BCAW-Tool/raw/master/Plots/All%20plots.jpg`

the current link is to https://github.com/AliYoussef96/BCAW-Tool/blob/master/Plots/All%20plots.jpg which is the github page containing the image, not the raw image itself

**Comment by AliYoussef96** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01500/joss.01500/10.21105.joss.01500.pdf)

**Comment by AliYoussef96** [comment]
@mgymrek I changed the path to the image in the .md, and it worked.

**Comment by AliYoussef96** [comment]
@mgymrek Can I ask what is next? are we ready to go further?

**Comment by mgymrek** [comment]
Thanks for fixing the proof. We are almost there. Last step: Can you now create a Zenodo archive and report the DOI in this thread. Please make sure the Zenodo archive has the same title and author list as the paper. Also please indicate whether the version needs to be updated (it is listed as v1.0.0 on this thread).

**Comment by AliYoussef96** [comment]
@whedon set v1.0.3 as version

**Comment by whedon** [comment]
I'm sorry @AliYoussef96, I'm afraid I can't do that. That's something only editors are allowed to do.

**Comment by AliYoussef96** [comment]
Hi @mgymrek, thank you for your response.

I created the Zenodo archive, DOI:  [![DOI](https://zenodo.org/badge/189912439.svg)](https://zenodo.org/badge/latestdoi/189912439) . With the same title and author list as the [paper](https://github.com/openjournals/joss-papers/blob/joss.01500/joss.01500/10.21105.joss.01500.pdf).



Yes, I want to update the version to v1.0.3 as pip [version](https://pypi.org/project/BCAWT/1.0.3/], I tried to do it, but I see only the editors are allowed to do that.

**Comment by Benjamin-Lee** [comment]
@arfon, who is responsible for copyediting the final paper?

**Comment by labarba** [comment]
@whedon set v1.0.3 as version

**Comment by whedon** [comment]
OK. v1.0.3 is the version.

**Comment by labarba** [comment]
@whedon set 10.5281/zenodo.3468557 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3468557" target="_blank">10.5281/zenodo.3468557</a> is the archive.

**Comment by labarba** [comment]
@mgymrek, @Benjamin-Lee — Are you ready to recommend publication of this paper?

**Comment by mgymrek** [comment]
@openjournals/joss-eics we are ready to accept this.

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Attempting to check references...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.1093/nar/15.3.1281 is OK
- 10.21105/joss.00905 is OK
- 10.1016/0378-1119(90)90491-9 is OK
- 10.1371/journal.pone.0194372 is OK
- 10.3389/fmicb.2017.01419 is OK
- 10.1371/journal.pgen.1000556 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01500/joss.01500/10.21105.joss.01500.pdf)

**Comment by Kevin-Mattheus-Moerman** [comment]
@AliYoussef96 some minor comments on the paper:

- [ ] Can  you remove the extra spaces introduced for items in brackets throughout the paper?
- [ ] Please amend your citation syntax for `...codons described by (Hersh-berg & Petrov, 2009) is implemented...` [See here](https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax). 
- [ ] Add hyphen to `built in` in: `...with built in and third-party packages...`
- [ ] Use `can` instead of `could` in "...tool could be divided into three groups."
- [ ] Should `BCAWT’s` be `BCAW’s`? in "...are reported in the BCAWT’s documentation."
- [ ] Consider rephrasing this sentence, it does not read well and may be confusing "The first one is data in CSV format described in (Table 1), the second group is plots summarized in (Fig 1), and the last one is text files each contains a different result for a different statistical test.". I propose (but correct it if I misunderstood): "The first group is data in the CSV format (see Table 1), the second group is plots (summarized in Fig 1), and the last group is text files, whereby each text file contains results for a different statistical test."
- [ ] Using 1-, 2- for a list in a sentence is unusual. Please change to: 1), 2). 
- [ ] Please rephrase this sentence which does not read well: `The advantages for BCAW tool over the existing tools are; 1- Automated Workflow, 2- No limitations on the number of genes, 3- The method used to determine optimal codons named correlation method is only available in the BCAW tool, 4- Plots as violin plots for nucleotides contents, and other plots are generated by the BCAW tool and no need for using another software to draw such plots.` I propose (but correct it if I misunderstood): `The advantages of BCAW tool over existing tools are; 1) the automated workflow, 2) the ability to handle large numbers of genes, 3) the method used to determine optimal codons, named the correlation method, is only available in the BCAW tool, 4) visualization and plotting capability, including the creation of violin plots for nucleotide contents, removing the need for other plotting software`. 
- [ ] Can you increase the fonts used in figure 1?
- [ ] Instead of `Further, ` can you sue `Furthermore, ` in "Further, using the correlation method..."
- [ ] You use both `the BCAW tool` and `BCAW tool`. Both read more awkward then BCAWT. I propose you use BCAWT throughout the paper instead of the aforementioned. 
- [ ] Add `a` in `BCAW tool returns CSV file containing...` Such that it reads: `BCAW tool returns a CSV file containing...`
- [ ] In `....containing the codon usage bias indices output ( Table 1 ).`, it seems you can use the acronym CUB instead. 
- [ ] In the table title: `Table 1: Explain the CSV files output...`, rephrase this to `Table 1: Explanation of the the CSV output file...`
- [ ] Please rephrase `Also, it returns 11 plot ( Fig 1 ), so one could interpret the results easily.` to `Furthermore, BCAWT returns 11 plots (Fig 1), enabling easy interpretation of the results.`
- [ ] Can you check the title for figure 1 `Fig 1: All output plots from BCAW tool analysis for coding sequence from Escherichia`, should that be "codon sequence"? or "a code sequence"? or "genetic code sequence"? 
- [ ] Please rephrase `The usage of BCAW Tool is made to be very easy where users have only to input a fasta format file containing genes to be analyzed, and variety of analyses will be performed.`
- [ ] Make sure you have an enter and spaces after the figure label to ensure the figure does not back the figure label (as it does not)

**Comment by AliYoussef96** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01500/joss.01500/10.21105.joss.01500.pdf)

**Comment by AliYoussef96** [comment]
- [x] Can you remove the extra spaces introduced for items in brackets throughout the paper?
- [x] Please amend your citation syntax for ...codons described by (Hersh-berg & Petrov, 2009)
- [x] Add hyphen to built in in: ...with built in and third-party packages...
- [x] Use can instead of could in "...tool could be divided into three groups."
- [x] Should BCAWT’s be BCAW’s? in "...are reported in the BCAWT’s documentation."
- [x] Consider rephrasing this sentence, it does not read well and may be confusing "The first one is data in CSV format described in (Table 1), the second group is plots summarized in (Fig 1), and the last one is text files each contains a different result for a different statistical test.". I propose (but correct it if I misunderstood): "The first group is data in the CSV format (see Table 1), the second group is plots (summarized in Fig 1), and the last group is text files, whereby each text file contains results for a different statistical test."
- [x] Using 1-, 2- for a list in a sentence is unusual. Please change to: 1), 2).
- [x] Please rephrase this sentence which does not read well: The advantages for BCAW tool over the existing tools are; 1- Automated Workflow, 2- No limitations on the number of genes, 3- The method used to determine optimal codons named correlation method is only available in the BCAW tool, 4- Plots as violin plots for nucleotides contents, and other plots are generated by the BCAW tool and no need for using another software to draw such plots. I propose (but correct it if I misunderstood): The advantages of BCAW tool over existing tools are; 1) the automated workflow, 2) the ability to handle large numbers of genes, 3) the method used to determine optimal codons, named the correlation method, is only available in the BCAW tool, 4) visualization and plotting capability, including the creation of violin plots for nucleotide contents, removing the need for other plotting software.
- [x] Can you increase the fonts used in figure 1? 
- [x] Instead of Further, can you sue Furthermore, in "Further, using the correlation method..."
- [x] You use both the BCAW tool and BCAW tool. Both read more awkward then BCAWT. I propose you use BCAWT throughout the paper instead of the aforementioned.
- [x] Add a in BCAW tool returns CSV file containing... Such that it reads: BCAW tool returns a CSV file containing...
- [x] In ....containing the codon usage bias indices output ( Table 1 )., it seems you can use the acronym CUB instead.
- [x] In the table title: Table 1: Explain the CSV files output..., rephrase this to Table 1: Explanation of the the CSV output file...
- [x] Please rephrase Also, it returns 11 plot ( Fig 1 ), so one could interpret the results easily. to Furthermore, BCAWT returns 11 plots (Fig 1), enabling easy interpretation of the results.
- [x] Can you check the title for figure 1 Fig 1: All output plots from BCAW tool analysis for coding sequence from Escherichia, should that be "codon sequence"? or "a code sequence"? or "genetic code sequence"?  >>> I meant ``` genes coding sequence```
- [x] Please rephrase The usage of BCAW Tool is made to be very easy where users have only to input a fasta format file containing genes to be analyzed, and variety of analyses will be performed.
- [x] Make sure you have an enter and spaces after the figure label to ensure the figure does not back the figure label (as it does not)

**Comment by Kevin-Mattheus-Moerman** [comment]
@AliYoussef96 thanks for implementing those. Some minor points remain: 

- [ ] Please use either `the BCAWT` throughout or `BCAWT`. I propose `BCAWT`, so to remove all instances of `the` in front of BCAWT. 
- [ ] Check again for BCAW and BCAWT (e.g. you still have `....are reported in the BCAW’s`)
- [ ] You currently do not have an acknowledgements section (which is optional), add one now if you feel it is appropriate. 
- [ ] You currently have two figure labels, i.e. above and below the figure. Can you use just one label under the figure? 
- [ ] Can you rephrase: `For example, to analyze genes coding sequence within Ecoli.fasta, save the results to a folder named save_path and choosing genetic code for bacteria:`, I propose (unless I misunderstood): ```For example, the following code snippet shows how to analyze genes for a coding sequence within the file `Ecoli.fasta`, for a genetic code specified, and to save the results to a folder named save_path.```

**Comment by AliYoussef96** [comment]
@Kevin-Mattheus-Moerman Thank you for your quick response!!

- [x] Please use either the BCAWT throughout or BCAWT. I propose BCAWT, so to remove all instances of the in front of BCAWT.
- [x] Check again for BCAW and BCAWT (e.g. you still have ....are reported in the BCAW’s)
- [x] You currently do not have an acknowledgements section (which is optional), add one now if you feel it is appropriate. >>> It is fine without.
- [x] You currently have two figure labels, i.e. above and below the figure. Can you use just one label under the figure?
- [x] Can you rephrase: For example, to analyze genes coding sequence within Ecoli.fasta, save the results to a folder named save_path and choosing genetic code for bacteria:, I propose (unless I misunderstood): For example, the following code snippet shows how to analyze genes for a coding sequence within the file `Ecoli.fasta`, for a genetic code specified, and to save the results to a folder named save_path.

**Comment by AliYoussef96** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01500/joss.01500/10.21105.joss.01500.pdf)

**Comment by Kevin-Mattheus-Moerman** [comment]
@AliYoussef96 getting close
- [ ] Add a full stop after `... a folder named save_path` so it reads `... a folder named save_path.`
- [ ] Remove the `the` in `...are reported in the BCAW’s` so it becomes: `.....are reported in BCAW’s`

**Comment by AliYoussef96** [comment]
• [x] Add a full stop after ... a folder named save_path so it reads ... a folder named save_path.
• [x] Remove the the in ...are reported in the BCAW’s so it becomes: .....are reported in BCAW’s

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01500/joss.01500/10.21105.joss.01500.pdf)

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.1093/nar/15.3.1281 is OK
- 10.21105/joss.00905 is OK
- 10.1016/0378-1119(90)90491-9 is OK
- 10.1371/journal.pone.0194372 is OK
- 10.3389/fmicb.2017.01419 is OK
- 10.1371/journal.pgen.1000556 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1035

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1035, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1185904964729790469) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1036
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.01500](https://doi.org/10.21105/joss.01500)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.01500/status.svg)](https://doi.org/10.21105/joss.01500)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.01500">
  <img src="https://joss.theoj.org/papers/10.21105/joss.01500/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.01500/status.svg
   :target: https://doi.org/10.21105/joss.01500
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.01500/status.svg)](https://doi.org/10.21105/joss.01500)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by AliYoussef96** [comment]
Huge thanks to @mgymrek, @Kevin-Mattheus-Moerman, and @Benjamin-Lee.  Your effort is greatly appreciated!
